File: node/strax_rpc_client.js

var PROTO_PATH = __dirname + '/../protos/strax_rpc.proto';

var async = require('async');
var grpc = require('grpc');
var protoLoader = require('@grpc/proto-loader');
var DataFrame = require('dataframe-js').DataFrame;
// Suggested options for similarity to existing grpc.load behavior
var packageDefinition = protoLoader.loadSync(
    PROTO_PATH,
    {keepCase: true,
     longs: String,
     enums: String,
     defaults: true,
     oneofs: true
    });
var protoDescriptor = grpc.loadPackageDefinition(packageDefinition);
// The protoDescriptor object has the full package hierarchy
var straxrpc = protoDescriptor.straxrpc;

var strax = new straxrpc.StraxRPC('localhost:50051',
                                       grpc.credentials.createInsecure());


/** 
* @param {function} callback Called when this demo is complete
*/

function runSearchField(pattern,callback) {
    // const pattern = "s1*";
    var spat = {
        pattern: pattern
    }
    console.log('\nLooking for fields that match the pattern '+ pattern+"...\n");
    var call = strax.searchField(spat);
    call.on('data', function(r) {
      console.log(r.name + " is part of " + r.data_name + " (provided by " + r.plugin + ")");
        });
    call.on('end', callback);
}

function runGetDF(run_id,dframe,callback) {
    // const run_id = "180423_1021";
    // const dframe = "event_basics";
    var call = strax.getDf({run_id:run_id, dframe:dframe})
    var data = {};
    var columns = [];
    call.on('data', function(r) {
        data[r.info.name] = r[r.info.dtype].values;
        columns.push(r.info.name);
      
        });
    
    call.on('end', function(){
        // console.log(columns);
        console.log("\n Got dataframe "+dframe+" back for run id "+run_id+"\n");
        var df = new DataFrame(data, columns);
        df.show();
        callback(df);
    }
     );
}


function main() {
    async.series([
        async.apply(runSearchField,"s1*"),
        async.apply(runGetDF,"180423_1021","event_basics"),
    ]);
  }
  
  if (require.main === module) {
    main();
  }
  
  exports.runSearchField = runSearchField;
  exports.runGetDF = runGetDF;
  
 

File: python/strax_rpc_client.py
from __future__ import print_function
import grpc
import strax_rpc_pb2, strax_rpc_pb2_grpc
import pandas as pd

def search_field(stub, pattern, max_matches=10):
    sp = strax_rpc_pb2.SearchPattern(
        pattern=pattern,
        max_matches=max_matches,
    )
    rs = stub.SearchField(sp)
    for r in rs:
        print(f"{r.name} is part of {r.data_name} (provided by {r.plugin})")

def data_info(stub, dataname):
    pi = strax_rpc_pb2.PluginInfo(name=dataname)
    # rs = stub.DataInfo(pi)
    data = {}
    for col in stub.DataInfo(pi):
        d = getattr(col, col.info.dtype)
        data[col.info.name] = pd.Series(index=d.index, data=d.values)
    df = pd.DataFrame(data)

    # t = PrettyTable()
    # t.field_names = ['index','Data name', 'Field name', 'Comment']
    # for i,r in enumerate(rs):
    #     t.add_row([i, r.name, r.dtype, r.comment ])
    print(df)

def get_df(stub,run_id,dframe):
    ti = strax_rpc_pb2.TableInfo(name=dframe, run_id=run_id)
    # f = strax_rpc_pb2.ColumnInfo(name='random')
    data = {}
    for col in stub.GetDF(ti):
        d = getattr(col, col.info.dtype)
        data[col.info.name] = pd.Series(index=d.index, data=d.values)
        # print(col.info.name, ' ', col.info.dtype)
        # for i,v in zip(r.index, r.value):
        #     print(f'{i} : {v}')
    df = pd.DataFrame(data)
    print(df)

def run():
    with grpc.insecure_channel('localhost:50051') as channel:
        stub = strax_rpc_pb2_grpc.StraxRPCStub(channel)
        print("---------------------- Search for s1* ----------------------")
        search_field(stub, "s1*")
        print("\n-------------- Data Info for event_basics --------------\n")
        data_info(stub, 'event_basics')
        print("\n----------- Get event_basics for 180423_1021 -----------\n")
        get_df(stub, '180423_1021', 'event_basics')


if __name__ == '__main__':
    run()
